Feature opt flush

bensander opened the pull request:
Passes all HC tests with HCC_OPT_FLUSH=0  or HCC_OPT_FLUSH=1

wsttiger committed:
Improve HCC_OPT_FLUSH

- Add marker if copy follows an AsyncOp without a completion_signal in same queue.
- Add system-scope release before D2H copy if queue _needs_sys_release
- Track _needs_sys_release when markers are enqueued.  This is more accurate
  tracking mechanism and avoids some unnecessary system-scope flushes.

bensander committed:
Increase LOOP_COUNT on async test so it runs longer.

Was finishing so fast the pending_counts were incorrect.

wsttiger committed:
Increase LOOP_COUNT on other tests to increase runtime.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -41,6 +41,7 @@
 #include <time.h>
 #include <iomanip>
 
+#define KALMAR_DEBUG (0)
```
@bensander can you remove this line?

whchung commented on include/kalmar_runtime.h:
```diff
@@ -245,7 +264,7 @@ class KalmarQueue
   /// enqueue marker with prior dependency
   virtual std::shared_ptr<KalmarAsyncOp> EnqueueMarkerWithDependency(int count, std::shared_ptr <KalmarAsyncOp> *depOps, memory_scope scope) { return nullptr; }
```
@bensander It seems the 3rd parameter, `scope` , is getting much more importance in this PR. Would you mind add more comments here to explain how `scope` works?

And looking at this PR I'm also wondering could this `scope` argument have a default value, say `hc::system_scope` ?

whchung commented on include/kalmar_runtime.h:
```diff
@@ -245,7 +264,7 @@ class KalmarQueue
   virtual std::shared_ptr<KalmarAsyncOp> EnqueueMarkerWithDependency(int count, std::shared_ptr <KalmarAsyncOp> *depOps, memory_scope scope) { return nullptr; }
 
-  virtual std::shared_ptr<KalmarAsyncOp> detectStreamDeps(hcCommandKind commandKind, KalmarAsyncOp *newCopyOp) { return nullptr; };
+  virtual std::shared_ptr<KalmarAsyncOp> detectStreamDeps(hcCommandKind commandKind, KalmarAsyncOp *newCopyOp, memory_scope *requiredRelease) { return nullptr; };
```
@bensander It seems the 3rd parameter, `requiredRelease`, is getting much more importance in this PR. Would you mind add more comments here to explain how the function works?

And would it be possible to change `requiredRelease` have be optional, like `memory_scope *requiredRelease = nullptr`? Because it seems to me not all call sites of `detectStreamDeps` really cares about the `memory_scope` changed by it.

whchung commented on include/hc.hpp:
```diff
@@ -1470,7 +1471,8 @@ inline completion_future accelerator_view::create_blocking_marker(completion_fut
     // This is necessary for example if copy command is followed by marker - we need the marker to wait for the copy to complete.
-    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr);
+    hc::memory_scope releaseScope = hc::no_scope;
+    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr, &releaseScope);
```
@bensander this is a place where we can consider make the 3rd parameter of `detectStreamDeps` be optional.

whchung commented on include/hc.hpp:
```diff
@@ -1493,7 +1495,8 @@ accelerator_view::create_blocking_marker(InputIterator first, InputIterator last
     // This is necessary for example if copy command is followed by marker - we need the marker to wait for the copy to complete.
-    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr);
+    hc::memory_scope releaseScope = hc::no_scope;
+    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr, &releaseScope);
```
@bensander this is a place where we can consider make the 3rd parameter of detectStreamDeps be optional.

whchung reviewed (changes requested):
@bensander This is an important PR. I'd like to propose a few changes before getting it merged. Besides more comments on the crucial APIs `detectStreamDeps` and `EnqueueMarkerWithDependency`, some arguments could also be considered made to be optional IMHO.

bensander committed:
Update releaseScope usage:

- Add documentation for EnqueueMarkerWithDependency and
  detectStreamDeps.
- Call greater_scope inside some create_marker functions missed in
  previous checkin.

bensander commented:
Jack – thanks for review!  I submitted an updated request.  The hc.hpp issues below should have called greater_scope -  I made that change.  Left releaseScope as a mandatory argument since this is internal function and we want called to explicitly and carefully choose their release scope.

From: "Wen-Heng (Jack) Chung" <notifications@github.com>
Reply-To: RadeonOpenCompute/hcc <reply@reply.github.com>
Date: Wednesday, May 24, 2017 at 9:47 AM
To: RadeonOpenCompute/hcc <hcc@noreply.github.com>
Cc: "Sander, Ben" <ben.sander@amd.com>, Mention <mention@noreply.github.com>
Subject: Re: [RadeonOpenCompute/hcc] Feature opt flush (#333)


@whchung requested changes on this pull request.

@bensander<https://github.com/bensander> This is an important PR. I'd like to propose a few changes before getting it merged. Besides more comments on the crucial APIs detectStreamDeps and EnqueueMarkerWithDependency, some arguments could also be considered made to be optional IMHO.

________________________________

In lib/hsa/mcwamp_hsa.cpp<https://github.com/RadeonOpenCompute/hcc/pull/333#discussion_r118269021>:

> @@ -41,6 +41,7 @@

 #include <time.h>

 #include <iomanip>



+#define KALMAR_DEBUG (0)

@bensander<https://github.com/bensander> can you remove this line?

________________________________

In include/kalmar_runtime.h<https://github.com/RadeonOpenCompute/hcc/pull/333#discussion_r118271215>:

> @@ -245,7 +264,7 @@ class KalmarQueue

   /// enqueue marker with prior dependency

   virtual std::shared_ptr<KalmarAsyncOp> EnqueueMarkerWithDependency(int count, std::shared_ptr <KalmarAsyncOp> *depOps, memory_scope scope) { return nullptr; }

@bensander<https://github.com/bensander> It seems the 3rd parameter, scope , is getting much more importance in this PR. Would you mind add more comments here to explain how scope works?

And looking at this PR I'm also wondering could this scope argument have a default value, say hc::system_scope ?

________________________________

In include/kalmar_runtime.h<https://github.com/RadeonOpenCompute/hcc/pull/333#discussion_r118271776>:

> @@ -245,7 +264,7 @@ class KalmarQueue

   /// enqueue marker with prior dependency

   virtual std::shared_ptr<KalmarAsyncOp> EnqueueMarkerWithDependency(int count, std::shared_ptr <KalmarAsyncOp> *depOps, memory_scope scope) { return nullptr; }



-  virtual std::shared_ptr<KalmarAsyncOp> detectStreamDeps(hcCommandKind commandKind, KalmarAsyncOp *newCopyOp) { return nullptr; };

+  virtual std::shared_ptr<KalmarAsyncOp> detectStreamDeps(hcCommandKind commandKind, KalmarAsyncOp *newCopyOp, memory_scope *requiredRelease) { return nullptr; };

@bensander<https://github.com/bensander> It seems the 3rd parameter, requiredRelease, is getting much more importance in this PR. Would you mind add more comments here to explain how the function works?

And would it be possible to change requiredRelease have be optional, like memory_scope *requiredRelease = nullptr? Because it seems to me not all call sites of detectStreamDeps really cares about the memory_scope changed by it.

________________________________

In include/hc.hpp<https://github.com/RadeonOpenCompute/hcc/pull/333#discussion_r118272765>:

> @@ -1470,7 +1471,8 @@ inline completion_future accelerator_view::create_blocking_marker(completion_fut



     // If necessary create an explicit dependency on previous command

     // This is necessary for example if copy command is followed by marker - we need the marker to wait for the copy to complete.

-    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr);

+    hc::memory_scope releaseScope = hc::no_scope;

+    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr, &releaseScope);

@bensander<https://github.com/bensander> this is a place where we can consider make the 3rd parameter of detectStreamDeps be optional.

________________________________

In include/hc.hpp<https://github.com/RadeonOpenCompute/hcc/pull/333#discussion_r118272880>:

> @@ -1493,7 +1495,8 @@ accelerator_view::create_blocking_marker(InputIterator first, InputIterator last



     // If necessary create an explicit dependency on previous command

     // This is necessary for example if copy command is followed by marker - we need the marker to wait for the copy to complete.

-    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr);

+    hc::memory_scope releaseScope = hc::no_scope;

+    std::shared_ptr<Kalmar::KalmarAsyncOp> depOp = pQueue->detectStreamDeps(hcCommandMarker, nullptr, &releaseScope);

@bensander<https://github.com/bensander> this is a place where we can consider make the 3rd parameter of detectStreamDeps be optional.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/RadeonOpenCompute/hcc/pull/333#pullrequestreview-40043930>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACYSAouofo3cozikBj5EliBQR-ZePvHCks5r9EMSgaJpZM4NkfbD>.

whchung reviewed (approved):
Thanks for the updated commit and clarifications. The PR now looks good to me.
